elastic/kibana
# [Search] Search Sessions with relative time range

## Summary

Part of https://github.com/elastic/kibana/issues/83640

In master background session restoration works only for absolute time ranges. 
This pr attempts to make it work for relative time ranges, by "syncing" Date.now() between different searches when relative time range is converted into absolute time range. 

Relative to absolute time conversion is not centralized even though we already had some `forceNow` logic for reporting. So some conversions could be missed and I am happy to follow on that

`NowProvider` is introduced in data services which keeps tracked of current `now` value. When session is started `sessionService` sets a new `now` into `NowProvider` so all subsequent searches using it for converting from relative to absolute time range: 
 
<details>
  <summary>high-level diagram</summary>
  
![Untitled-2020-12-07-1233-3](https://user-images.githubusercontent.com/7784120/103633760-72f69600-4f46-11eb-8d72-8ba20f44fcee.png)
</details>

In future instead of using `now` for converting just time filter we would explore using [point at time api](https://www.elastic.co/guide/en/elasticsearch/reference/master/point-in-time-api.html)

There was an edge case with **auto refresh**. Previously new session wasn't created on auto refresh fetches and with new approach it would cause the same `now` for all refreshes. **Now new sessions are created on auto refresh fetches. For this autoRefresh subscription was moved on a dashboard level and removed from embeddable layer**. https://github.com/elastic/kibana/pull/84405/#discussion_r542285700

### How to test

```
xpack.data_enhanced.search.sendToBackground.enabled: true
data.search.aggs.shardDelay.enabled: true # helpful to create "slow" visualizations
```

1. Create a dashboard with search, visualize and lens embeddable. 
2. Make a search with relative time range. Save it. 
3. Get `searchSessionId` from any of inspector. Get absolute time range from any of inspectors.
4. build a dashboard URL to restore a session: `${dashboardURL}&searchSessionId=${searchSessionId}` and replace relative time range with copied time range
5. Open it. Search session should be restored and you should see absolute time range

**^^ this is what functional test doest**


### Demo 

I took this pr + https://github.com/elastic/kibana/pull/81707 to record this demo: 

https://user-images.githubusercontent.com/7784120/103648107-a0e6d500-4f5c-11eb-957d-93a2a6f61dc7.mov

Use this branch to try it yourself: https://github.com/Dosant/kibana/tree/feature/data/management-ui-plus

### Checklist


### For maintainers
